payments/api.py

```diff
@@ -0,0 +1,405 @@
+    pending orders that are in the system.
+
+    Args:
+    - refnos (list or None): check specific reference numbers
+    Returns:
+    - Tuple of counts: fulfilled count, cancelled count, error count
+
+    """
+
+    gateway = PaymentGateway.get_gateway_class(
+        settings.ECOMMERCE_DEFAULT_PAYMENT_GATEWAY
+    )
+
+    if refnos is not None:
+        pending_orders = PendingOrder.objects.filter(
+            state=PendingOrder.STATE.PENDING, reference_number__in=refnos
+        ).values_list("reference_number", flat=True)
+    else:
+        pending_orders = PendingOrder.objects.filter(
+            state=PendingOrder.STATE.PENDING
+        ).values_list("reference_number", flat=True)
+
+    if len(pending_orders) == 0:
+        return (0, 0, 0)
+
+    msg = f"Resolving {len(pending_orders)} orders"
+    log.info(msg)
+
+    results = gateway.find_and_get_transactions(pending_orders)
+
+    if len(results.keys()) == 0:
+        msg = "No orders found to resolve."
+        log.info(msg)
+        return (0, 0, 0)
+
+    fulfilled_count = cancel_count = error_count = 0
+
+    for result in results:
+        payload = results[result]
+        if int(payload["reason_code"]) == CYBERSOURCE_REASON_CODE_SUCCESS:
+            try:
+                order = PendingOrder.objects.filter(
+                    state=PendingOrder.STATE.PENDING,
+                    reference_number=payload["req_reference_number"],
+                ).get()
+
+                order.fulfill(payload)
+                order.save()
+                fulfilled_count += 1
+
+                msg = f"Fulfilled order {order.reference_number}."
+                log.info(msg)
+            except Exception as e:
+                msg = (
+                    "Couldn't process pending order for fulfillment "
+                    f"{payload['req_reference_number']}: {e!s}"
+                )
+                log.exception(msg)
+                error_count += 1
+        else:
+            try:
+                order = PendingOrder.objects.filter(
+                    state=PendingOrder.STATE.PENDING,
+                    reference_number=payload["req_reference_number"],
+                ).get()
+
+                order.cancel()
+                order.transactions.create(
+                    transaction_id=payload["transaction_id"],
+                    amount=order.total_price_paid,
+                    data=payload,
+                    reason=f"Cancelled due to processor code {payload['reason_code']}",
+                )
+                order.save()
+                cancel_count += 1
+
+                msg = f"Cancelled order {order.reference_number}."
+                log.info(msg)
+            except Exception as e:
+                msg = (
+                    "Couldn't process pending order for cancellation "
+                    f"{payload['req_reference_number']}: {e!s}"
+                )
+                log.exception(msg)
```

Comment:
Exception logging should use `%s` + args instead of f-strings.

Reply:
fixed in https://github.com/mitodl/unified-ecommerce/pull/99/commits/e5185ab87cc8c8f933cb3b8ddd79d05e497ef77a